internal/services/cosmos/cosmosdb_mongo_user_definition_resource.go

```diff
@@ -0,0 +1,274 @@
+			if err != nil {
+				return err
+			}
+
+			mongoUserDefinitionId := fmt.Sprintf("%s.%s", databaseId.Name, model.Username)
+			id := mongorbacs.NewMongodbUserDefinitionID(databaseId.SubscriptionId, databaseId.ResourceGroup, databaseId.DatabaseAccountName, mongoUserDefinitionId)
+
+			locks.ByName(id.DatabaseAccountName, CosmosDbAccountResourceName)
+			defer locks.UnlockByName(id.DatabaseAccountName, CosmosDbAccountResourceName)
+
+			existing, err := client.MongoDBResourcesGetMongoUserDefinition(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for existing %s: %+v", id, err)
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			properties := mongorbacs.MongoUserDefinitionCreateUpdateParameters{
+				Properties: &mongorbacs.MongoUserDefinitionResource{
+					DatabaseName: pointer.To(databaseId.Name),
+					Mechanisms:   pointer.To("SCRAM-SHA-256"),
+					Password:     pointer.To(model.Password),
+					UserName:     pointer.To(model.Username),
+					Roles:        expandInheritedRole(model.InheritedRoleNames, databaseId.Name),
+				},
+			}
+
+			if err := client.MongoDBResourcesCreateUpdateMongoUserDefinitionThenPoll(ctx, id, properties); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r CosmosDbMongoUserDefinitionResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Cosmos.MongoRBACClient
+
+			id, err := mongorbacs.ParseMongodbUserDefinitionID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			locks.ByName(id.DatabaseAccountName, CosmosDbAccountResourceName)
+			defer locks.UnlockByName(id.DatabaseAccountName, CosmosDbAccountResourceName)
+
+			var model CosmosDbMongoUserDefinitionResourceModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			resp, err := client.MongoDBResourcesGetMongoUserDefinition(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			properties := resp.Model
+			if properties == nil {
+				return fmt.Errorf("retrieving %s: properties was nil", id)
+			}
+
+			databaseId, err := parse.MongodbDatabaseID(model.CosmosMongoDatabaseId)
+			if err != nil {
+				return err
+			}
+
+			parameters := mongorbacs.MongoUserDefinitionCreateUpdateParameters{
+				Properties: &mongorbacs.MongoUserDefinitionResource{
+					DatabaseName: pointer.To(databaseId.Name),
+					Mechanisms:   pointer.To("SCRAM-SHA-256"),
+					Password:     pointer.To(model.Password),
+					UserName:     pointer.To(model.Username),
+					Roles:        expandInheritedRole(model.InheritedRoleNames, databaseId.Name),
+				},
+			}
```

Comment:
could we use `metadata.ResourceData.HasChange` when updating these properties while using the values returned from the Get request to populate any which do not have changes?

Reply:
Updated